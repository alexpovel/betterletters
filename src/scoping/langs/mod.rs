use super::Scoper;
#[cfg(doc)]
use crate::scoping::scope::Scope::{In, Out};
use crate::{find::Find, ranges::Ranges};
use log::{debug, trace};
use std::str::FromStr;
pub use tree_sitter::{
    Language as TSLanguage, Parser as TSParser, Query as TSQuery, QueryCursor as TSQueryCursor,
};

/// C#.
pub mod csharp;
/// Go.
pub mod go;
/// Hashicorp Configuration Language
pub mod hcl;
/// Python.
pub mod python;
/// Rust.
pub mod rust;
/// TypeScript.
pub mod typescript;

/// Represents a (programming) language.
#[derive(Debug)]
pub struct Language<Q> {
    query: Q,
}

impl<Q> Language<Q> {
    /// Create a new language with the given associated query over it.
    pub fn new(query: Q) -> Self {
        Self { query }
    }
}

/// A query over a language, for scoping.
///
/// Parts hit by the query are [`In`] scope, parts not hit are [`Out`] of scope.
#[derive(Debug, Clone)]
pub enum CodeQuery<C, P>
where
    C: FromStr + Into<TSQuery>,
    P: Into<TSQuery>,
{
    /// A custom, user-defined query.
    Custom(C),
    /// A prepared query.
    ///
    /// Availability depends on the language, respective languages features, and
    /// implementation in this crate.
    Prepared(P),
}

impl<C, P> From<CodeQuery<C, P>> for TSQuery
where
    C: FromStr + Into<TSQuery>,
    P: Into<TSQuery>,
{
    fn from(value: CodeQuery<C, P>) -> Self {
        match value {
            CodeQuery::Custom(query) => query.into(),
            CodeQuery::Prepared(query) => query.into(),
        }
    }
}

/// In a query, use this name to mark a capture to be ignored.
///
/// Useful for queries where tree-sitter doesn't natively support a fitting node type,
/// and a result is instead obtained by ignoring unwanted parts of bigger captures.
pub(super) const IGNORE: &str = "_SRGN_IGNORE";

/// A scoper for a language.
///
/// Functions much the same, but provides specific language-related functionality.
pub trait LanguageScoper: Scoper + Find {
    /// The language's tree-sitter language.
    fn lang() -> TSLanguage
    where
        Self: Sized; // Exclude from trait object

    /// The language's tree-sitter query.
    fn query(&self) -> TSQuery
    where
        Self: Sized; // Exclude from trait object

    /// The language's tree-sitter parser.
    #[must_use]
    fn parser() -> TSParser
    where
        Self: Sized, // Exclude from trait object
    {
        let mut parser = TSParser::new();
        parser
            .set_language(&Self::lang())
            .expect("Should be able to load language grammar and parser");

        parser
    }

    /// Scope the given input using the language's query.
    ///
    /// In principle, this is the same as [`Scoper::scope`].
    fn scope_via_query(query: &mut TSQuery, input: &str) -> Ranges<usize>
    where
        Self: Sized, // Exclude from trait object
    {
        // tree-sitter is about incremental parsing, which we don't use here
        let old_tree = None;

        trace!("Parsing into AST: {:?}", input);

        let tree = Self::parser()
            .parse(input, old_tree)
            .expect("No language set in parser, or other unrecoverable error");

        let root = tree.root_node();
        debug!(
            "S expression of parsed source code is: {:?}",
            root.to_sexp()
        );

        let run = |query: &mut TSQuery| {
            trace!("Running query: {:?}", query);

            let mut qc = TSQueryCursor::new();
            let matches = qc.matches(query, root, input.as_bytes());

            let mut ranges: Ranges<usize> = matches
                .flat_map(|query_match| query_match.captures)
                .map(|capture| capture.node.byte_range())
                .collect();

            // ⚠️ tree-sitter queries with multiple captures will return them in some
            // mixed order (not ordered, and not merged), but we later rely on cleanly
            // ordered, non-overlapping ranges (a bit unfortunate we have to know about
            // that remote part over here).
            ranges.merge();
            trace!("Querying yielded ranges: {:?}", ranges);

            ranges
        };

        let ranges = run(query);

        let is_ignored = |name: &str| name.starts_with(IGNORE);
        let has_ignored_captures = query.capture_names().iter().any(|name| is_ignored(name));

        if has_ignored_captures {
            let ignored_ranges = {
                let acknowledged_captures = query
                    .capture_names()
                    .iter()
                    .filter(|name| !is_ignored(name))
                    .map(|s| String::from(*s))
                    .collect::<Vec<_>>();

                for name in acknowledged_captures {
                    trace!("Disabling capture for: {:?}", name);
                    query.disable_capture(&name);
                }

                debug!("Query has captures to ignore: running additional query");
                run(query)
            };

            let res = ranges - ignored_ranges;
            debug!("Ranges cleaned up after subtracting ignores: {:?}", res);

            res
        } else {
            ranges
        }
    }
}
