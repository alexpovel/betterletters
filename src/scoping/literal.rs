use super::{scope::RangesWithContext, Scoper};
use crate::ranges::Ranges;
use log::trace;
use std::{error::Error, fmt, ops::Range};
use unescape::unescape;

/// A literal string for querying.
#[derive(Debug)]
pub struct Literal(String);

/// An error that can occur when parsing a literal.
#[derive(Debug)]
pub enum LiteralError {
    /// The literal contains invalid escape sequences.
    InvalidEscapeSequences(String),
}

impl fmt::Display for LiteralError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::InvalidEscapeSequences(literal) => {
                write!(f, "Contains invalid escape sequences: '{literal}'")
            }
        }
    }
}

impl Error for LiteralError {}

impl TryFrom<String> for Literal {
    type Error = LiteralError;

    fn try_from(literal: String) -> Result<Self, Self::Error> {
        let unescaped =
            unescape(&literal).ok_or(LiteralError::InvalidEscapeSequences(literal.to_string()))?;

        Ok(Self(unescaped))
    }
}

impl Scoper for Literal {
    fn scope_raw<'viewee>(&self, input: &'viewee str) -> RangesWithContext<'viewee> {
        let ranges = {
            let len = self.0.len();

            let ranges: Ranges<usize> = input
                .match_indices(&self.0)
                .map(|(i, _)| Range {
                    start: i,
                    end: i + len,
                })
                .collect();

            trace!("Ranges in scope for {:?}: {:?}", self, ranges);

            ranges
        };

        ranges.into()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::scoping::{
        scope::{
            RWScope, RWScopes,
            Scope::{In, Out},
        },
        view::ScopedView,
    };
    use rstest::rstest;
    use std::borrow::Cow::Borrowed;

    #[rstest]
    #[case(
        "a",
        "a",
        ScopedView::new(
            RWScopes(vec![
                RWScope(In(Borrowed("a"), None)),
            ])
        )
    )]
    #[case(
        "aa",
        "a",
        ScopedView::new(
            RWScopes(vec![
                RWScope(In(Borrowed("a"), None)),
                RWScope(In(Borrowed("a"), None)),
            ])
        )
    )]
    #[case(
        "aba",
        "a",
        ScopedView::new(
            RWScopes(vec![
                RWScope(In(Borrowed("a"), None)),
                RWScope(Out("b")),
                RWScope(In(Borrowed("a"), None)),
            ])
        )
    )]
    //
    #[case(
        ".",
        ".",
        ScopedView::new(
            RWScopes(vec![
                RWScope(In(Borrowed("."), None)),
            ])
        )
    )]
    #[case(
        r"\.",
        ".",
        ScopedView::new(
            RWScopes(vec![
                RWScope(Out(r"\")),
                RWScope(In(Borrowed("."), None)),
            ])
        )
    )]
    #[case(
        r".",
        r"\\.",
        ScopedView::new(
            RWScopes(vec![
                RWScope(Out(r".")),
            ])
        )
    )]
    //
    #[case(
        "Hello\nWorld\n",
        "\n",
        ScopedView::new(
            RWScopes(vec![
                RWScope(Out("Hello")),
                RWScope(In(Borrowed("\n"), None)),
                RWScope(Out("World")),
                RWScope(In(Borrowed("\n"), None)),
            ])
        )
    )]
    fn test_literal_scoping(
        #[case] input: &str,
        #[case] literal: &str,
        #[case] expected: ScopedView,
    ) {
        let mut builder = crate::scoping::view::ScopedViewBuilder::new(input);
        let literal = Literal::try_from(literal.to_owned()).unwrap();
        builder.explode(&literal);
        let actual = builder.build();

        assert_eq!(actual, expected);
    }
}
