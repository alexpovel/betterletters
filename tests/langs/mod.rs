mod csharp;
mod go;
mod hcl;
mod rust;
mod typescript;

use rstest::rstest;
use serde::{Deserialize, Serialize};
use srgn::scoping::{
    langs::{
        python::{PreparedPythonQuery, Python},
        CodeQuery, LanguageScoper,
    },
    regex::Regex,
    scope::Scope,
    view::ScopedViewBuilder,
};
use std::{fs::read_to_string, ops::Range, path::Path};

fn get_input_output(lang: &str, file: &str) -> (String, String) {
    let path = Path::new("tests/langs");
    let path = path.join(lang);

    let input = read_to_string(path.join(format!("in/{file}"))).unwrap();
    let output = read_to_string(path.join(format!("out/{file}"))).unwrap();

    (input, output)
}

/// Nuke the target character from the input.
///
/// Convenience function for testing, as deleting a specific character, while
/// *retaining* it elsewhere, where the language did *not* scope down, is an easy way to
/// test.
fn nuke_target(input: &str, lang: &impl LanguageScoper) -> String {
    let mut builder = ScopedViewBuilder::new(input);

    builder.explode(lang);

    // Needs to be ASCII such that we can target e.g. variable names.
    let target = String::from("__T__");
    builder.explode(&Regex::try_from(target).unwrap());

    let mut view = builder.build();
    view.delete();

    view.to_string()
}

/// A type that when serialized, will visually highlight the portions of a line which
/// were matched.
///
/// Single letter names to not disrupt alignment.
///
/// For example, when serialized as YAML, produces output such as
///
/// ```yaml
/// n: 156
/// l: "        print(f\"Loop iteration {i}\")\n"
/// m: "        ^^^^^                           "
/// ```
///
/// with `m` indicating the [`Scope::In`] part of the `l`ine.
#[derive(Debug, Serialize, Deserialize)]
struct InScopeLinePart {
    /// Line number.
    n: usize,
    /// Line itself, in original form.
    l: String,
    /// The string highlighting the matching. Has to be serialized *below*.
    m: String,
}

impl InScopeLinePart {
    fn new(line_number: usize, line_contents: String, match_: String, span: Range<usize>) -> Self {
        // Split into the three components
        let (start, mid, end) = (
            &line_contents[..span.start],
            &line_contents[span.clone()],
            &line_contents[span.end..],
        );

        // ASSUMPTION is that `.escape_default()` will escape the same way (YAML, ...)
        // serialization does. Important for alignment to be correct.

        // Leading space
        let mut m = " ".repeat(start.escape_default().to_string().len());
        // The highlights for the line above.
        m.push_str(&"^".repeat(mid.escape_default().to_string().len()));
        // Trailing spaces; not strictly necessary but slightly nicer.
        m.push_str(&" ".repeat(end.escape_default().to_string().len()));

        assert_eq!(
            line_contents[span], match_,
            "What is highlighted is what was matched"
        );

        Self {
            n: line_number,
            l: line_contents,
            m,
        }
    }
}

#[rstest]
#[case(
    "base.py_comments",
    include_str!("python/base.py"),
    Python::new(CodeQuery::Prepared(PreparedPythonQuery::Comments)),
)]
#[case(
    "base.py_strings",
    include_str!("python/base.py"),
    Python::new(CodeQuery::Prepared(PreparedPythonQuery::Strings)),
)]
#[case(
    "base.py_imports",
    include_str!("python/base.py"),
    Python::new(CodeQuery::Prepared(PreparedPythonQuery::Imports)),
)]
#[case(
    "base.py_docstrings",
    include_str!("python/base.py"),
    Python::new(CodeQuery::Prepared(PreparedPythonQuery::DocStrings)),
)]
#[case(
    "base.py_function-names",
    include_str!("python/base.py"),
    Python::new(CodeQuery::Prepared(PreparedPythonQuery::FunctionNames)),
)]
#[case(
    "base.py_function-calls",
    include_str!("python/base.py"),
    Python::new(CodeQuery::Prepared(PreparedPythonQuery::FunctionCalls)),
)]
#[case(
    "base.py_class",
    include_str!("python/base.py"),
    Python::new(CodeQuery::Prepared(PreparedPythonQuery::Class)),
)]
fn test_language_scopers(
    #[case] snapshot_name: &str,
    #[case] contents: &str,
    #[case] lang: impl LanguageScoper,
) {
    let mut builder = ScopedViewBuilder::new(contents);
    builder.explode(&lang);
    let view = builder.build();

    // Collect only those lines which are in scope. This avoids enormous clutter from
    // out of scope items. The in-scope lines are *pinned* by lines numbers and
    // highlights/their span, so false-positives and confusion around multiple matches
    // per line are avoided.
    let inscope_parts: Vec<InScopeLinePart> = view
        .lines()
        .into_iter()
        .enumerate()
        .flat_map(|(i, line)| {
            let mut start = 0;

            line.scopes()
                .0
                .clone()
                .into_iter()
                .filter_map(move |scope| {
                    let contents: &str = (&scope).into();
                    let end = start + contents.len();
                    let span = start..end;
                    let part =
                        InScopeLinePart::new(i + 1, line.to_string(), contents.to_string(), span);
                    start = end;

                    match scope.0 {
                        Scope::In(..) => Some(part),
                        Scope::Out(..) => None,
                    }
                })
        })
        .collect();

    insta::assert_yaml_snapshot!(snapshot_name, inscope_parts);
}
