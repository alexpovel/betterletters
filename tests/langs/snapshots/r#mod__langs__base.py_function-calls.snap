---
source: tests/langs/mod.rs
expression: inscope_parts
---
- n: 24
  l: "    print(f\"Global test_var is now {test_var}\")\n"
  m: "    ^^^^^                                          "
- n: 32
  l: "        print(\"Function decorator called\")\n"
  m: "        ^^^^^                                 "
- n: 33
  l: "        return func(*args, **kwargs)\n"
  m: "               ^^^^                   "
- n: 41
  l: "    print(\"Inside decorated function\")\n"
  m: "    ^^^^^                                 "
- n: 56
  l: "            print(\"Static method decorator called\")\n"
  m: "            ^^^^^                                      "
- n: 57
  l: "            return func(*args, **kwargs)\n"
  m: "                   ^^^^                   "
- n: 66
  l: "        print(f\"Class variable is now {cls.class_var}\")\n"
  m: "        ^^^^^                                              "
- n: 72
  l: "        print(f\"Instance variable is {self.instance_var}\")\n"
  m: "        ^^^^^                                                 "
- n: 78
  l: "        print(\"Inside static method\")\n"
  m: "        ^^^^^                            "
- n: 102
  l: "squared_numbers = [\"x\" + square(x) for x in range(10)]\n"
  m: "                           ^^^^^^                         "
- n: 102
  l: "squared_numbers = [\"x\" + square(x) for x in range(10)]\n"
  m: "                                              ^^^^^       "
- n: 105
  l: "unique_squares = {square(x) for x in range(10)}\n"
  m: "                  ^^^^^^                         "
- n: 105
  l: "unique_squares = {square(x) for x in range(10)}\n"
  m: "                                     ^^^^^       "
- n: 108
  l: "squares_dict = {x: square(x) for x in range(10)}\n"
  m: "                   ^^^^^^                         "
- n: 108
  l: "squares_dict = {x: square(x) for x in range(10)}\n"
  m: "                                      ^^^^^       "
- n: 116
  l: "            raise ValueError(\"Negative value\")\n"
  m: "                  ^^^^^^^^^^                      "
- n: 118
  l: "            raise ZeroDivisionError(\"Division by zero\")\n"
  m: "                  ^^^^^^^^^^^^^^^^^                        "
- n: 121
  l: "        print(f\"Caught an exception: {e}\")\n"
  m: "        ^^^^^                                 "
- n: 123
  l: "        print(f\"Caught an exception: {e}\")\n"
  m: "        ^^^^^                                 "
- n: 125
  l: "        print(\"No exceptions caught\")\n"
  m: "        ^^^^^                            "
- n: 127
  l: "        print(\"This will always be printed\")\n"
  m: "        ^^^^^                                   "
- n: 139
  l: "    inner()\n"
  m: "    ^^^^^    "
- n: 140
  l: "    print(f\"Nonlocal variable is {nonlocal_var}\")\n"
  m: "    ^^^^^                                            "
- n: 150
  l: "    print(f\"Inplace operations result: {x}\")\n"
  m: "    ^^^^^                                       "
- n: 158
  l: "        print(\"test_var is greater than 5\")\n"
  m: "        ^^^^^                                  "
- n: 160
  l: "        print(\"test_var is 5 or less\")\n"
  m: "        ^^^^^                             "
- n: 165
  l: "        print(f\"Counter is {counter}\")\n"
  m: "        ^^^^^                             "
- n: 169
  l: "    for i in range(3):\n"
  m: "             ^^^^^      "
- n: 170
  l: "        print(f\"Loop iteration {i}\")\n"
  m: "        ^^^^^                           "
- n: 173
  l: "    with open(__file__) as f:\n"
  m: "         ^^^^                  "
- n: 175
  l: "        print(\"Read from file:\", content)\n"
  m: "        ^^^^^                                "
- n: 183
  l: "            print(\"Zero\")\n"
  m: "            ^^^^^            "
- n: 185
  l: "            print(\"One\")\n"
  m: "            ^^^^^           "
- n: 187
  l: "            print(\"Other\")\n"
  m: "            ^^^^^             "
- n: 194
  l: "    print(\"Async function executed\")\n"
  m: "    ^^^^^                               "
- n: 199
  l: "    free_func()\n"
  m: "    ^^^^^^^^^    "
- n: 200
  l: "    decorated_func()\n"
  m: "    ^^^^^^^^^^^^^^    "
- n: 202
  l: "    instance = TestClass()\n"
  m: "               ^^^^^^^^^    "
- n: 205
  l: "    print(square(5))\n"
  m: "    ^^^^^             "
- n: 205
  l: "    print(square(5))\n"
  m: "          ^^^^^^      "
- n: 206
  l: "    exception_handling(0)\n"
  m: "    ^^^^^^^^^^^^^^^^^^     "
- n: 207
  l: "    modify_nonlocal()\n"
  m: "    ^^^^^^^^^^^^^^^    "
- n: 208
  l: "    inplace_operations()\n"
  m: "    ^^^^^^^^^^^^^^^^^^    "
- n: 209
  l: "    control_flow()\n"
  m: "    ^^^^^^^^^^^^    "
- n: 210
  l: "    match_statement(1)\n"
  m: "    ^^^^^^^^^^^^^^^     "
- n: 211
  l: "    asyncio.run(async_function())\n"
  m: "                ^^^^^^^^^^^^^^     "
